x-pack/filebeat/input/awss3/s3_test.go

```diff
@@ -0,0 +1,261 @@
+		ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+		defer cancel()
+
+		ctrl, ctx := gomock.WithContext(ctx, t)
+		defer ctrl.Finish()
+		mockS3API := NewMockS3API(ctrl)
+		mockPublisher := NewMockBeatClient(ctrl)
+
+		s3Event := newS3Event("log.txt")
+
+		mockS3API.EXPECT().
+			GetObject(gomock.Any(), gomock.Eq(s3Event.S3.Bucket.Name), gomock.Eq(s3Event.S3.Object.Key)).
+			Return(nil, errFakeConnectivityFailure)
+
+		s3ObjProc := newS3ObjectProcessorFactory(logp.NewLogger(inputName), nil, mockS3API, mockPublisher, nil)
+		ack := newEventACKTracker(ctx)
+		err := s3ObjProc.Create(ctx, logp.NewLogger(inputName), ack, s3Event).ProcessS3Object()
+		require.Error(t, err)
+		assert.True(t, errors.Is(err, errFakeConnectivityFailure), "expected errFakeConnectivityFailure error")
+	})
+}
+
+func testProcessS3Object(t testing.TB, file, contentType string, numEvents int, selectors ...fileSelectorConfig) []beat.Event {
+	return _testProcessS3Object(t, file, contentType, numEvents, false, selectors)
+}
+
+func testProcessS3ObjectError(t testing.TB, file, contentType string, numEvents int, selectors ...fileSelectorConfig) []beat.Event {
+	return _testProcessS3Object(t, file, contentType, numEvents, true, selectors)
+}
+
+func _testProcessS3Object(t testing.TB, file, contentType string, numEvents int, expectErr bool, selectors []fileSelectorConfig) []beat.Event {
+	t.Helper()
+
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	ctrl, ctx := gomock.WithContext(ctx, t)
+	defer ctrl.Finish()
+	mockS3API := NewMockS3API(ctrl)
+	mockPublisher := NewMockBeatClient(ctrl)
+
+	s3Event, s3Resp := newS3Object(t, file, contentType)
+	var events []beat.Event
+	gomock.InOrder(
+		mockS3API.EXPECT().
+			GetObject(gomock.Any(), gomock.Eq(s3Event.S3.Bucket.Name), gomock.Eq(s3Event.S3.Object.Key)).
+			Return(s3Resp, nil),
+		mockPublisher.EXPECT().
+			Publish(gomock.Any()).
+			Do(func(event beat.Event) { events = append(events, event) }).
+			Times(numEvents),
+	)
+
+	s3ObjProc := newS3ObjectProcessorFactory(logp.NewLogger(inputName), nil, mockS3API, mockPublisher, selectors)
+	ack := newEventACKTracker(ctx)
+	err := s3ObjProc.Create(ctx, logp.NewLogger(inputName), ack, s3Event).ProcessS3Object()
+
+	if !expectErr {
+		require.NoError(t, err)
+		assert.Equal(t, numEvents, len(events))
+		assert.EqualValues(t, numEvents, ack.pendingACKs)
+	} else {
+		require.Error(t, err)
+	}
+
+	return events
+}
+
+// TestNewMockS3Pager verifies that newMockS3Pager is behaving similar to
+// the AWS S3 Pager.
+func TestNewMockS3Pager(t *testing.T) {
+	fakeObjects := []s3.Object{
+		{Key: awssdk.String("foo")},
+		{Key: awssdk.String("bar")},
+		{Key: awssdk.String("baz")},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	ctrl, ctx := gomock.WithContext(ctx, t)
+	defer ctrl.Finish()
+	mockS3Pager := newMockS3Pager(ctrl, 1, fakeObjects)
```

Comment:
@aspacca I update the S3 interface to account for your need to call ListObjects. I included a helper to mock the S3 pagination calls for your tests.